### PR TITLE
Add Unreachability tests

### LINKED ISSUE #34630: Write Tests for Unreachability
No of people needed : 1

- [x] Write as much as possible to cover all cases and send a PR
- [x] Report bugs if any 

Reported bugs:
-   #34914
-   #34820
-   #34734
-   #34722
-   #34707

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$title

Fixes #34630

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/reachability/ReachabilityAnalysisTest.java [hunk 1]
```diff
@@ -86,6 +86,33 @@ public void testReachability(String function) {
         };
     }
 
+    @Test(dataProvider = "reachabilityTests2")
+    public void testReachability2(String function) {
+        result = BCompileUtil.compile("test-src/reachability-analysis/reachability_analysis2.bal");
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/reachability/ReachabilityAnalysisTest.java:91]: Other than the current bal file becoming larger, is there a reason to introduce these in a separate file anyway? Are the tests significantly different?

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal [hunk 2]
```diff
@@ -0,0 +1,1034 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+function executeTestReachabilityWithIfContainingReturn() {
+    assertEqual(testReachabilityWithIfContainingReturn(), "Ballerina");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal:18]: If one of these fail the rest won't run, right? We used to write tests this way and switch to call individual functions using data providers (from the Java side) due to the same reason.

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal [hunk 3]
```diff
@@ -0,0 +1,1034 @@
...
+    assertEqual(testReachabilityWithIfContainingReturn2(), "Ballerina");
+    assertEqual(testReachabilityWithIfContainingReturn3(), "Hello");
+    assertEqual(testReachabilityWithIfContainingReturn4(), "Hello");
+    assertEqual(testReachabilityWithIfContainingReturn5(), "Ballerina");
+    assertEqual(testReachabilityWithIfContainingReturn6(), "Ballerina");
+    assertEqual(testReachabilityWithIfContainingReturn7(), "Ballerina");
+}
+
+function testReachabilityWithIfContainingReturn() returns string {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal:27]: Usual pattern would be to have something like `reachabilityWithIfContainingReturn` be the name of the function with this logic and then `testReachabilityWithIfContainingReturn` calling `reachabilityWithIfContainingReturn` and doing the assertion.

Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal:27]: IMO `reachabilityWithIfContainingReturn` does not imply action. Every function is an action, so the name should contain at least one verb right?

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal (hunk 2 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal:18]: Will update

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/reachability/ReachabilityAnalysisTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/reachability/ReachabilityAnalysisTest.java:91]: There can be duplicated tests also if we merge.

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal (hunk 3 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/reachability_analysis2.bal:27]: What I meant to say was that the "test" prefix should be for the function in which the assertion is done.

**Review by MaryamZi — COMMENTED** [review]
Changes LGTM, need to finalize the branch.

**Review by MaryamZi — APPROVED** [review]

**Review by hasithaa — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/unreachability_test2.bal [hunk 4]
```diff
@@ -0,0 +1,1219 @@
...
+        return;
+    }
+
+    panic unionResult;
+}
+
+public function testUnreachabilityWithQueryAction3() {
+    string[] stringArray = ["Hello", " ", "World"];
+
+    error? unionResult = from var item in stringArray
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/unreachability_test2.bal:1100]: Shall we also add test case similar to

```ballerina
function testUnreachableStatementInQueryAction() {
    string m;
    error(m) = <error>from var item in 1 ... 5
        where true
        do {
            while true {
                int _ = 3;
            }
            int _ = 2; //Unreachable
        };
}
```

**Pull request merged**